<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <meta http-equiv="Content-Style-Type" content="text/css">
  <title>CCP4 Program Suite</title>
  <meta name="Generator" content="Cocoa HTML Writer">
  <meta name="CocoaVersion" content="949.54">
  <style type="text/css">
    p.p1 {margin: 0.0px 0.0px 16.0px 0.0px; font: 24.0px Times}
    p.p2 {margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times}
    p.p3 {margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times}
    p.p4 {margin: 0.0px 0.0px 12.0px 0.0px; font: 12.0px Times}
    p.p5 {margin: 0.0px 0.0px 12.0px 0.0px; font: 12.0px Times; min-height: 14.0px}
    p.p6 {margin: 0.0px 0.0px 0.0px 36.0px; text-indent: -36.0px; font: 12.0px Times; color: #0023ee; min-height: 14.0px}
    p.p7 {margin: 0.0px 0.0px 0.0px 36.0px; text-indent: -36.0px; font: 16.0px Times}
    p.p8 {margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times; min-height: 14.0px}
    p.p9 {margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times}
    p.p10 {margin: 0.0px 0.0px 14.0px 0.0px; font: 12.0px Times}
    p.p11 {margin: 0.0px 0.0px 0.0px 36.0px; text-indent: -36.0px; font: 12.0px Times}
    span.s1 {text-decoration: underline ; color: #0023ee}
    span.s2 {font: 12.0px Lucida Grande}
    span.Apple-tab-span {white-space:pre}
    ol.ol1 {list-style-type: decimal}
    ul.ul1 {list-style-type: disc}
  </style>
</head>
<body>
<h1 style="margin: 0.0px 0.0px 16.0px 0.0px; font: 24.0px Times"><b>AMPLE (CCP4: Unsupported Program)</b></h1>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>NAME</b></h2>
<p class="p3"><b>AMPLE </b>- automated de novo search model generation for molecular replacement</p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>SYNOPSIS</b></h2>
<p class="p4">Utilises programs like Rosetta to generate ab initio models to be used in molecular replacement:</p>
<p class="p4"><b>HAMMER_Beta_Release.py</b> [-h] [-ROSETTA <i>ROSETTA_path</i>] [-RDB <i>ROSETTA_database</i>] [-fragsexe <i>path to make_fragments.pl</i>] [-Rosetta_cluster <i>path to Rosettas cluster</i>] [-fasta <i>fasta_file</i>] [-name <i>priotein name</i>] [-NProc <i>NoProcessors</i>] [-RunDir <i>run_directory</i>] [-SCWRL <i>path to scwrl</i>] [-LGA <i>path_to_LGA dir</i>] [-MAX <i>Maxcluster exe</i>] [-THESEUS <i>Theseus exe required</i>] [-MTZ <i>MTZ in</i>] [-MODELS <i>folder of decoys</i>] [-MakeModels <i>Do the modelling</i>] [-ROSETTA_DIR <i>Rosetta_dir</i>] [<a href="file:///Users/jac45/Downloads/hammer_doc.html#options"><span class="s1">Command line options</span></a>]</p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>DESCRIPTION</b></h2>
<p class="p4">AMPLE has three main parts:</p>
<ol class="ol1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">For a given target sequence, calls Rosetta for the ab initio modeling of protein decoys, or reads in a set of pre-made models</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Preparation of decoys into search ensembles for molecular replacement</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Running molecular replacement using these search ensembles and inputting MR results into SHELX</li>
</ol>
<p class="p5"><br></p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>DEPENDENCIES</b></h2>
<p class="p4">Before AMPLE can be used, the following dependencies should be installed on the local system.</p>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times"><b>SCWRL</b> (version 4.0) available from: <a href="http://dunbrack.fccc.edu/scwrl4/#installation"><span class="s1">SCWRL</span></a></li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times"><b>LGA</b> (Local-Global Alignment)(ver. 10/2009) only available by contacting the author <a href="http://proteinmodel.org/AS2TS/system.html"><span class="s1">LGA</span></a></li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times"><b>Rosetta</b> (tested with versions 3.2 and 3.3) <a href="http://www.rosettacommons.org/"><span class="s1">Rosetta</span></a></li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times"><b>Theseus</b> (ver. 1.4.3) <a href="http://www.theseus3d.org/src/theseus_align"><span class="s1">Theseus</span></a></li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times"><b>Maxcluster</b> (ver 0.6.6) (as a replacement for LGA) <a href="http://www.sbg.bio.ic.ac.uk/%7Emaxcluster/download.html"><span class="s1">Maxcluster</span></a></li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times"><b>SHELX </b>(Beta version of shelxe)</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">phenix_MR_Rosetta (optional)</li>
</ul>
<p class="p6"><br></p>
<p class="p7"><b>Editing the Setup File</b></p>
<p class="p3">AMPLE comes with a setup file that needs to be modified with the paths of the dependancies:</p>
<p class="p8"><br></p>
<p class="p3">export LGA_PATH=&lt;path to LGA install directory&gt;</p>
<p class="p3">export THESEUS_EXE=&lt;path to theseus executable&gt;</p>
<p class="p3">export SCWRL_EXE=&lt;path to SCWRL executable&gt;</p>
<p class="p3">export PYTHONPATH=$PYTHONPATH:$CCP4/share/ample/using</p>
<p class="p3">export ROSETTA_PATH=&lt;path to Rosetta install directory&gt;</p>
<p class="p8"><br></p>
<p class="p3">Also needs the shelx paths</p>
<p class="p8"><br></p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>INPUT AND OUTPUT FILES</b></h2>
<h3 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times"><b>HKLIN</b></h3>
<p class="p3">Input structure factor file for target structure. Must include a FreeR_flag column.</p>
<p class="p8"><br></p>
<p class="p8"><br></p>
<p class="p9"><b>Intermediate Output files</b></p>
<p class="p3">Generated models will output to the "Models" directory.</p>
<p class="p3">The models are clustered and the clustered models are output to the "clusters" directory.</p>
<p class="p3">Ensembles are made and places in the "ensembles" directory.<span class="Apple-converted-space"> </span></p>
<p class="p3">MRBUMP will run in the "MRBUMP" directory.</p>
<p class="p3">SHELX will run in each MRBUMP run directory</p>
<p class="p8"><br></p>
<p class="p9"><b>Final Output files</b></p>
<p class="p10">Final outputs will be placed in the "RESULTS" directory</p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>COMMAND LINE INPUT</b></h2>
<p class="p4">optional arguments:</p>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-h, --help     show this help message and exit</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-ROSETTA <i>ROSETTA_path</i>     path for Rosetta AbinitioRelax</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-RDB <i>ROSETTA_database</i>     path for Rosetta database</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-fragsexe <i>path to make_fragments.pl</i>     location of make_fragments.pl</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-Rosetta_cluster <i>path to Rosettas cluster</i>     location of rosetta cluster</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-fasta <i>fasta_file</i>     protein fasta file. (required)</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-name <i>priotein name</i>     name of protien in the format ABCD</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-NProc <i>NoProcessors</i>     number of processers (default 1)</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-RunDir <i>run_directory</i>     directory to put files (default current dir)</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-SCWRL <i>path to scwrl</i>     pathway to SCWRL exe</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-LGA <i>path_to_LGA dir</i>     pathway to LGA folder (not the exe) will use the 'lga' executable</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-MAX <i>Maxcluster exe</i>     Maxcluster exe</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-THESEUS <i>Theseus exe (required)</i>     Theseus exe</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-MTZ <i>MTZ in</i>     MTZ in</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-MODELS <i>folder of decoys</i>     folder of decoys</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-MakeModels <i>Do the modelling</i>     run rosetta modeling, set to False to import pre-made models (required if making models locally default True)</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-ROSETTA_DIR <i>Rosetta_dir</i>     the Rosetta install directory</li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-F <i>Flag for F</i></li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-SIGF <i>Flag for SIGF</i></li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-FREE <i>FreeR flag</i></li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-CLUSTER True if using a cluster, false if a multi-core PC<span class="Apple-converted-space"> </span></li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-percent percent of the model to truncate at each interval<span class="Apple-converted-space"> </span></li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-NumShelxCyc number of shelxe cycles to perform<span class="Apple-converted-space">?| </span></li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-NoShelx will not run shelx, will only run MRBUMP<span class="Apple-converted-space">?| </span></li>
</ul>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">-TRYALL if True will try all ensembles, if False will exit at the first success<span class="Apple-converted-space">?| </span></li>
</ul>




<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>Using the GUI</b></h2>
<h3 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times"><b>Input Files</b></h3>
<p class="p4">HKLIN: mtz file. The flags F, SIGF and FREE must be set.<span class="s2"><br>
</span>number of processors: int for parallel running for speed<span class="s2"><br>
</span>Fasta: must have a new line char at end of sequence<span class="s2"><br>
</span>work directory: where all run files are generated</p>
<h3 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times"><b>Models Import</b></h3>
<p class="p4">Make Models: if you want Rosetta to make the models locally, specify "True". If you have models already made specify "False" and give a path to a folder containing them.<span class="s2"><br>
</span>Models Path: specify "none" to make them locally. give a path if importing.</p>
<h3 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times"><b>Rosetta Quick</b></h3>
<p class="p4">Rosetta Install Directory: Install folder of Rosetta will try to find paths of the executables.</p>
<h3 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times"><b>Rosetta Full Pathways</b></h3>
<p class="p4">for a non standard install, the paths may be different. So give the following paths:<span class="s2"><br>
</span>AbinitioRelax.linuxgccrelease (executable)<span class="s2"><br>
</span>rosetta_database (folder pathway)<span class="s2"><br>
</span>make_fragments.pl (executable)<span class="s2"><br>
</span>cluster.linuxgccrelease (executable)</p>
<h3 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times"><b>3rd Party Locations</b></h3>
<p class="p4">For some functions some 3rd party executables are needed, give the pathways (if not set by setup.sh):<span class="s2"><br>
</span>maxcluster (executable)<span class="s2"><br>
</span>Scwrl4 (executable)<span class="s2"><br>
</span>theseus (executable)<span class="s2"><br>
</span>LGA (install folder)</p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>Known Issues:</b></h2>
<p class="p5"><br></p>
<ol class="ol1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">The output names of the Rosetta files are hard coded, if the Rosetta naming conventions change this needs to be updated.</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Assumes that the number of models to make is &gt; number of processors</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Crash handling by making models in different dirs, can rerun the job and Rosetta will continue from where it stopped. But if the folders are already full of models, Rosetta will hang- advise rerunning in a different folder and imputing if modelling complete.</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Theseus is also known to hang (will be caught and rerun until success)</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Rosetta cluster can fail (will be caught and rerun until success)</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Assumes paths are correct, checks for pathways to be added</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Needs a long command line (a problem with theseus)</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Shelx will not run if the spacegroup is P-1, this is changed to P 1</li>
</ol>
<p class="p5"><br></p>
<p class="p4">The final version will include:</p>
<ul class="ul1">
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">user feedback about performance of modelling</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">on the fly clustering/truncation to optimise successes</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">a normal mode (to remove run files) and a debug mode (all files kept)</li>
  <li style="margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Times">Missing Domain modeling</li>
</ul>
<p class="p5"><br></p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>EXAMPLE USAGE</b></h2>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>PROGRAM OUTPUT</b></h2>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>AUTHORS</b></h2>
<p class="p3">Jaclyn Bibby, University of Liverpool, UK<span class="Apple-converted-space"> </span></p>
<h2 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 18.0px Times"><b>ACKNOWLEDGEMENTS</b></h2>
<p class="p4">Daniel Rigden, Martyn Winn, Olga Mayans.</p>
<h3 style="margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Times"><b>AMPLE Program References</b></h3>
<p class="p4">Any publication arising from use of AMPLE should include the following reference:</p>
<p class="p4">???????????????????</p>
<p class="p3">In addition, authors of specific programs should be referenced where applicable:</p>
<p class="p3">MrBUMP</p>
<p class="p3">R.M.Keegan and M.D.Winn (2007) <i>Acta Cryst.</i> <b>D63</b>, 447-457</p>
<p class="p3">CCP4</p>
<p class="p3">Collaborative Computational Project, Number 4. (1994), "The CCP4 Suite: Programs for Protein Crystallography". <i>Acta Cryst.</i> <b>D50</b>, 760-763</p>
<p class="p3">FASTA</p>
<p class="p3">W. R. Pearson and D. J. Lipman (1988), "Improved Tools for Biological Sequence Analysis", <i>PNAS</i> <b>85</b>, 2444-2448</p>
<p class="p3">MOLREP</p>
<p class="p3">A.A.Vagin &amp; A.Teplyakov (1997) <i>J. Appl. Cryst.</i> <b>30</b>, 1022-1025</p>
<p class="p3">PHASER</p>
<p class="p3">McCoy, A.J., Grosse-Kunstleve, R.W., Storoni, L.C. &amp; Read, R.J. (2005). "Likelihood-enhanced fast translation functions" <i>Acta Cryst</i> <b>D61</b>, 458-464</p>
<p class="p3">REFMAC</p>
<p class="p3">G.N. Murshudov, A.A.Vagin and E.J.Dodson, (1997) "Refinement of Macromolecular Structures by the Maximum-Likelihood Method" <i>Acta Cryst.</i> <b>D53</b>, 240-255</p>
</body>
</html>
